#
#	修改：      bionic/libc/Android.mk
#	修改：      bionic/libc/SYSCALLS.TXT
#	修改：      bionic/libc/arch-arm/syscalls.mk
#	重命名：    bionic/libc/arch-arm/syscalls/epoll_create.S -> bionic/libc/arch-arm/syscalls/epoll_create1.S
#	修改：      bionic/libc/arch-mips/syscalls.mk
#	重命名：    bionic/libc/arch-mips/syscalls/epoll_create.S -> bionic/libc/arch-mips/syscalls/epoll_create1.S
#	修改：      bionic/libc/arch-x86/syscalls.mk
#	重命名：    bionic/libc/arch-x86/syscalls/epoll_create.S -> bionic/libc/arch-x86/syscalls/epoll_create1.S
#	新文件：    bionic/libc/bionic/epoll_create.cpp
#	修改：      bionic/libc/include/sys/epoll.h
#	修改：      bionic/libc/include/sys/socket.h
#	修改：      bionic/linker/linker.cpp
#
diff --git a/bionic/libc/Android.mk b/bionic/libc/Android.mk
index 1b02b78..0901464 100644
--- a/bionic/libc/Android.mk
+++ b/bionic/libc/Android.mk
@@ -205,6 +205,7 @@ libc_bionic_src_files := \
     bionic/brk.cpp \
     bionic/dirent.cpp \
     bionic/__errno.c \
+    bionic/epoll_create.cpp \
     bionic/eventfd_read.cpp \
     bionic/eventfd_write.cpp \
     bionic/futimens.cpp \
diff --git a/bionic/libc/SYSCALLS.TXT b/bionic/libc/SYSCALLS.TXT
index 88c980f..ce6cb8a 100644
--- a/bionic/libc/SYSCALLS.TXT
+++ b/bionic/libc/SYSCALLS.TXT
@@ -305,7 +305,7 @@ long    perf_event_open(struct perf_event_attr *attr_uptr, pid_t pid, int cpu, i
 int	futex(void *, int, int, void *, void *, int) 1
 
 # epoll
-int     epoll_create(int size)     1
+int     epoll_create1(int size)     1
 int     epoll_ctl(int epfd, int op, int fd, struct epoll_event *event)    1
 int     epoll_wait(int epfd, struct epoll_event *events, int max, int timeout)   1
 
diff --git a/bionic/libc/arch-arm/syscalls.mk b/bionic/libc/arch-arm/syscalls.mk
index 252a428..e17705b 100644
--- a/bionic/libc/arch-arm/syscalls.mk
+++ b/bionic/libc/arch-arm/syscalls.mk
@@ -196,7 +196,7 @@ syscall_src += arch-arm/syscalls/sysinfo.S
 syscall_src += arch-arm/syscalls/personality.S
 syscall_src += arch-arm/syscalls/perf_event_open.S
 syscall_src += arch-arm/syscalls/futex.S
-syscall_src += arch-arm/syscalls/epoll_create.S
+syscall_src += arch-arm/syscalls/epoll_create1.S
 syscall_src += arch-arm/syscalls/epoll_ctl.S
 syscall_src += arch-arm/syscalls/epoll_wait.S
 syscall_src += arch-arm/syscalls/inotify_init.S
diff --git a/bionic/libc/arch-arm/syscalls/epoll_create.S b/bionic/libc/arch-arm/syscalls/epoll_create.S
deleted file mode 100644
index b2894e6..0000000
--- a/bionic/libc/arch-arm/syscalls/epoll_create.S
+++ /dev/null
@@ -1,15 +0,0 @@
-/* autogenerated by gensyscalls.py */
-#include <asm/unistd.h>
-#include <linux/err.h>
-#include <machine/asm.h>
-
-ENTRY(epoll_create)
-    mov     ip, r7
-    ldr     r7, =__NR_epoll_create
-    swi     #0
-    mov     r7, ip
-    cmn     r0, #(MAX_ERRNO + 1)
-    bxls    lr
-    neg     r0, r0
-    b       __set_errno
-END(epoll_create)
diff --git a/bionic/libc/arch-arm/syscalls/epoll_create1.S b/bionic/libc/arch-arm/syscalls/epoll_create1.S
new file mode 100644
index 0000000..d05451d
--- /dev/null
+++ b/bionic/libc/arch-arm/syscalls/epoll_create1.S
@@ -0,0 +1,15 @@
+/* autogenerated by gensyscalls.py */
+#include <asm/unistd.h>
+#include <linux/err.h>
+#include <machine/asm.h>
+
+ENTRY(epoll_create1)
+    mov     ip, r7
+    ldr     r7, =__NR_epoll_create1
+    swi     #0
+    mov     r7, ip
+    cmn     r0, #(MAX_ERRNO + 1)
+    bxls    lr
+    neg     r0, r0
+    b       __set_errno
+END(epoll_create1)
diff --git a/bionic/libc/arch-mips/syscalls.mk b/bionic/libc/arch-mips/syscalls.mk
index 23393a2..994f114 100644
--- a/bionic/libc/arch-mips/syscalls.mk
+++ b/bionic/libc/arch-mips/syscalls.mk
@@ -199,7 +199,7 @@ syscall_src += arch-mips/syscalls/sysinfo.S
 syscall_src += arch-mips/syscalls/personality.S
 syscall_src += arch-mips/syscalls/perf_event_open.S
 syscall_src += arch-mips/syscalls/futex.S
-syscall_src += arch-mips/syscalls/epoll_create.S
+syscall_src += arch-mips/syscalls/epoll_create1.S
 syscall_src += arch-mips/syscalls/epoll_ctl.S
 syscall_src += arch-mips/syscalls/epoll_wait.S
 syscall_src += arch-mips/syscalls/inotify_init.S
diff --git a/bionic/libc/arch-mips/syscalls/epoll_create.S b/bionic/libc/arch-mips/syscalls/epoll_create.S
deleted file mode 100644
index 2f4ac7f..0000000
--- a/bionic/libc/arch-mips/syscalls/epoll_create.S
+++ /dev/null
@@ -1,22 +0,0 @@
-/* autogenerated by gensyscalls.py */
-#include <asm/unistd.h>
-    .text
-    .globl epoll_create
-    .align 4
-    .ent epoll_create
-
-epoll_create:
-    .set noreorder
-    .cpload $t9
-    li $v0, __NR_epoll_create
-    syscall
-    bnez $a3, 1f
-    move $a0, $v0
-    j $ra
-    nop
-1:
-    la $t9,__set_errno
-    j $t9
-    nop
-    .set reorder
-    .end epoll_create
diff --git a/bionic/libc/arch-mips/syscalls/epoll_create1.S b/bionic/libc/arch-mips/syscalls/epoll_create1.S
new file mode 100644
index 0000000..a13b495
--- /dev/null
+++ b/bionic/libc/arch-mips/syscalls/epoll_create1.S
@@ -0,0 +1,22 @@
+/* autogenerated by gensyscalls.py */
+#include <asm/unistd.h>
+    .text
+    .globl epoll_create1
+    .align 4
+    .ent epoll_create1
+
+epoll_create1:
+    .set noreorder
+    .cpload $t9
+    li $v0, __NR_epoll_create1
+    syscall
+    bnez $a3, 1f
+    move $a0, $v0
+    j $ra
+    nop
+1:
+    la $t9,__set_errno
+    j $t9
+    nop
+    .set reorder
+    .end epoll_create1
diff --git a/bionic/libc/arch-x86/syscalls.mk b/bionic/libc/arch-x86/syscalls.mk
index 11573de..beb58bc 100644
--- a/bionic/libc/arch-x86/syscalls.mk
+++ b/bionic/libc/arch-x86/syscalls.mk
@@ -200,7 +200,7 @@ syscall_src += arch-x86/syscalls/sysinfo.S
 syscall_src += arch-x86/syscalls/personality.S
 syscall_src += arch-x86/syscalls/perf_event_open.S
 syscall_src += arch-x86/syscalls/futex.S
-syscall_src += arch-x86/syscalls/epoll_create.S
+syscall_src += arch-x86/syscalls/epoll_create1.S
 syscall_src += arch-x86/syscalls/epoll_ctl.S
 syscall_src += arch-x86/syscalls/epoll_wait.S
 syscall_src += arch-x86/syscalls/inotify_init.S
diff --git a/bionic/libc/arch-x86/syscalls/epoll_create.S b/bionic/libc/arch-x86/syscalls/epoll_create.S
deleted file mode 100644
index 061f173..0000000
--- a/bionic/libc/arch-x86/syscalls/epoll_create.S
+++ /dev/null
@@ -1,21 +0,0 @@
-/* autogenerated by gensyscalls.py */
-#include <linux/err.h>
-#include <machine/asm.h>
-#include <asm/unistd.h>
-
-ENTRY(epoll_create)
-    pushl   %ebx
-    mov     8(%esp), %ebx
-    movl    $__NR_epoll_create, %eax
-    int     $0x80
-    cmpl    $-MAX_ERRNO, %eax
-    jb      1f
-    negl    %eax
-    pushl   %eax
-    call    __set_errno
-    addl    $4, %esp
-    orl     $-1, %eax
-1:
-    popl    %ebx
-    ret
-END(epoll_create)
diff --git a/bionic/libc/arch-x86/syscalls/epoll_create1.S b/bionic/libc/arch-x86/syscalls/epoll_create1.S
new file mode 100644
index 0000000..5972d49
--- /dev/null
+++ b/bionic/libc/arch-x86/syscalls/epoll_create1.S
@@ -0,0 +1,21 @@
+/* autogenerated by gensyscalls.py */
+#include <linux/err.h>
+#include <machine/asm.h>
+#include <asm/unistd.h>
+
+ENTRY(epoll_create1)
+    pushl   %ebx
+    mov     8(%esp), %ebx
+    movl    $__NR_epoll_create1, %eax
+    int     $0x80
+    cmpl    $-MAX_ERRNO, %eax
+    jb      1f
+    negl    %eax
+    pushl   %eax
+    call    __set_errno
+    addl    $4, %esp
+    orl     $-1, %eax
+1:
+    popl    %ebx
+    ret
+END(epoll_create1)
diff --git a/bionic/libc/bionic/epoll_create.cpp b/bionic/libc/bionic/epoll_create.cpp
new file mode 100644
index 0000000..1dfafa8
--- /dev/null
+++ b/bionic/libc/bionic/epoll_create.cpp
@@ -0,0 +1,33 @@
+/*
+ * Copyright (C) 2013 The Android Open Source Project
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *  * Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ *  * Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
+ * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ */
+
+#include <sys/epoll.h>
+
+int epoll_create(int /*obsolete_size*/) {
+  return epoll_create1(0);
+}
diff --git a/bionic/libc/include/sys/epoll.h b/bionic/libc/include/sys/epoll.h
index 38739aa..4e2b787 100644
--- a/bionic/libc/include/sys/epoll.h
+++ b/bionic/libc/include/sys/epoll.h
@@ -29,6 +29,7 @@
 #define _SYS_EPOLL_H_
 
 #include <sys/cdefs.h>
+#include <asm/fcntl.h> /* For O_CLOEXEC. */
 
 __BEGIN_DECLS
 
@@ -51,6 +52,8 @@ __BEGIN_DECLS
 #define EPOLL_CTL_DEL    2
 #define EPOLL_CTL_MOD    3
 
+#define EPOLL_CLOEXEC O_CLOEXEC
+
 typedef union epoll_data 
 {
     void *ptr;
@@ -66,6 +69,7 @@ struct epoll_event
 };
 
 int epoll_create(int size);
+int epoll_create1(int flags);
 int epoll_ctl(int epfd, int op, int fd, struct epoll_event *event);
 int epoll_wait(int epfd, struct epoll_event *events, int max, int timeout);
 
diff --git a/bionic/libc/include/sys/socket.h b/bionic/libc/include/sys/socket.h
index 17ba0a1..21be130 100644
--- a/bionic/libc/include/sys/socket.h
+++ b/bionic/libc/include/sys/socket.h
@@ -32,6 +32,7 @@
 #include <sys/types.h>
 #include <linux/socket.h>
 
+#include <asm/fcntl.h>
 #include <asm/socket.h>
 #include <linux/sockios.h>
 #include <linux/uio.h>
@@ -61,13 +62,15 @@ typedef int socklen_t;
 #define SOCK_PACKET      10
 #endif
 
-/* BIONIC: second argument to shutdown() */
+#define SOCK_CLOEXEC O_CLOEXEC
+#define SOCK_NONBLOCK O_NONBLOCK
+
 enum {
-    SHUT_RD = 0,        /* no more receptions */
+  SHUT_RD = 0,
 #define SHUT_RD         SHUT_RD
-    SHUT_WR,            /* no more transmissions */
+  SHUT_WR,
 #define SHUT_WR         SHUT_WR
-    SHUT_RDWR           /* no more receptions or transmissions */
+  SHUT_RDWR
 #define SHUT_RDWR       SHUT_RDWR
 };
 
diff --git a/bionic/linker/linker.cpp b/bionic/linker/linker.cpp
index d776bdf..dc282f1 100644
--- a/bionic/linker/linker.cpp
+++ b/bionic/linker/linker.cpp
@@ -1531,8 +1531,8 @@ static bool soinfo_link_image(soinfo* si) {
          * phdr_table_protect_segments() after all of them are applied
          * and all constructors are run.
          */
-        DEBUG("%s has text relocations. This is wasting memory and is "
-                "a security risk. Please fix.", si->name);
+        //DEBUG("%s has text relocations. This is wasting memory and is "
+         //       "a security risk. Please fix.", si->name);
         if (phdr_table_unprotect_segments(si->phdr, si->phnum, si->load_bias) < 0) {
             DL_ERR("can't unprotect loadable segments for \"%s\": %s",
                    si->name, strerror(errno));
